templates/scripts/util.js

```diff
@@ -0,0 +1,142 @@
+/**
+ * Acquires all data from the given form and POSTs it as JSON to the target URL.
+ * In case of failure this function will throw an error.
+ * In case of success a parsed JSON body of the response will be returned,
+ * unless the body contains a `location` field,
+ * in that case the page will be redirected to that location.
+ *
+ * @param formId - ID of the form.
+ * @param target - Target URL to POST to. Defaults to the current URL.
+ * @returns {Promise<unknown>} - The response JSON.
+ */
+async function postJsonForm(formId, target = '') {
+  const form = document.getElementById(formId);
+  const formData = new FormData(form);
+  const res = await fetch(target, {
+    method: 'POST',
+    credentials: 'include',
+    headers: { 'accept': 'application/json', 'content-type': 'application/json' },
+    body: JSON.stringify(Object.fromEntries(formData)),
+  });
+  if (res.status >= 400) {
+    const error = await res.json();
+    throw new Error(`${error.statusCode} - ${error.name}: ${error.message}`)
+  } else if (res.status === 200 || res.status === 201) {
+    const body = await res.json();
+    if (body.location) {
+      location.href = body.location;
+    } else {
+      return body;
+    }
+  }
+}
+
+/**
+ * Redirects the page to the given target with the key/value pairs of the JSON body as query parameters.
+ * Controls will be deleted from the JSON to prevent very large URLs.
+ * `false` values will be deleted to prevent incorrect serializations to "false".
+ * @param json - JSON to convert.
+ * @param target - URL to redirect to.
+ */
+function redirectJsonResponse(json, target) {
+  // These would cause the URL to get very large, can be acquired later if needed
+  delete json.controls;
+
+  // Remove false parameters since these would be converted to "false" strings
+  for (const [key, val] of Object.entries(json)) {
+    if (typeof val === 'boolean' && !val) {
+      delete json[key];
+    }
+  }
+
+  const searchParams = new URLSearchParams(Object.entries(json));
+  location.href = `${target}?${searchParams.toString()}`;
+}
+
+/**
+ * Adds a listener to the given form to catch the form submission and do an API call instead.
+ * In case of an error, the inner text of the given error block will be updated with the message.
+ * In case of success the callback function will be called.
+ *
+ * @param formId - ID of the form.
+ * @param errorId - ID of the error block.
+ * @param apiTarget - Target URL to send the POST request to. Defaults to the current URL.
+ * @param callback - Callback function that will be called with the response JSON.
+ */
+function addPostListener(formId, errorId, apiTarget, callback) {
```

Comment:
Instead of callback, return a (failing) promise.

Reply:
I don't understand how this would work. This function adds a listener to the form. Do you mean that his function should return a Promise containing the result of posting the form? This would not work if the first post returns an error and only the second post succeeds right?

Reply:
Ah it's repeated, understood.